File: 400.nth-digit.py
#
# @lc app=leetcode id=400 lang=python3
#
# [400] Nth Digit
#

# Time:  O(logn)
# Space: O(1)

# @lc code=start
class Solution:
    def findNthDigit(self, n: int) -> int:
        digit_len, count, start = 1, 9, 1
        while n > digit_len * count:
            n -= digit_len * count
            digit_len += 1
            count *= 10
            start *= 10
        start += (n - 1) // digit_len
        return int(str(start)[(n - 1) % digit_len])


# @lc code=end

File: 400.nth-digit.js
/*
 * @lc app=leetcode id=400 lang=javascript
 *
 * [400] Nth Digit
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var findNthDigit = function (n) {
  let digitLen = 1;
  let count = 9;
  let start = 1;
  while (digitLen * count < n) {
    n -= digitLen * count;
    digitLen += 1;
    count *= 10;
    start *= 10;
  }
  start += Math.trunc((n - 1) / digitLen);
  return Number(`${start}`[(n - 1) % digitLen]);
};
// @lc code=end
